staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -601,6 +690,53 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
 	return info.storages[info.storageVersion].CustomResource, nil
 }
 
+// getOrCreateServingInfoForWrite is getOrCreateServingInfoFor, but blocks CR
+// write requests until the storage version update is processed, aborted, or
+// timed-out.
+// If the storage version update is processed, unblock and allow the write request.
+// If the storage version update is aborted, unblock and reject the write request.
+// If the storage version update timed-out, unblock and allow the write request.
+//
+// If aborted is true, the returned error (403) should be surfaced to the client,
+// indicating that the server forbids the CR request due to the CRD deletion.
+// Otherwise, any non-nil error returned by getOrCreateServingInfoFor may be
+// wrapped by the server as an internal error (500) before surfacing to the client.
+func (r *crdHandler) getOrCreateServingInfoForWrite(uid types.UID, name string) (ret *crdInfo, err error, aborted bool) {
+	ret, err = r.getOrCreateServingInfoFor(uid, name)
+
+	// Make sure the StorageVersionAPI feature gate is enabled.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) ||
+		ret.storageVersionUpdateProcessedCh == nil {
+		return ret, err, false
+	}
+
+	// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+	// to register themselves to the wait group. Ideally the storage version update should
+	// not cause the requests to miss the 1s window; otherwise the requests may
+	// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+	// first CR request establishes the underlying storage).
+	select {
+	case <-ret.storageVersionUpdateProcessedCh:
+	case <-ret.storageVersionUpdateAbortedCh:
+		// Storage version update failed because the CRD was deleted. Here
+		// we return a 403 to fail this in-flight request.
+		return nil, apierrors.NewForbidden(schema.GroupResource{
```

Comment:
the forbidden error is misplaced here. Would rather return `nil, nil, true` here, and react to `abort==true` in the handler. The handler knows which error to return to the caller (403).